plugins/pulp_python/plugins/models.py

```diff
@@ -1,120 +1,150 @@
-                             will represent.
-        :type  archive_path: basestring
-        :return:             An instance of Package that represents the package found at
-                             archive_path.
-        :rtype:              pulp_python.plugins.models.Package
-        :raises:             ValueError if archive_path does not point to a valid Python tarball
-                             created with setup.py sdist.
-        :raises:             IOError if the archive_path does not exist.
-        """
-        try:
-            compression_type = cls._compression_type(archive_path)
-            checksum = cls.checksum(archive_path)
-            package_archive = tarfile.open(archive_path)
-            metadata_file = None
-            for member in package_archive.getmembers():
-                if re.match('.*/PKG-INFO$|^PKG-INFO$', member.name):
-                    # find the metadata file with the shortest path
-                    if metadata_file:
-                        if len(member.name) < len(metadata_file.name):
-                            metadata_file = member
-                    else:
-                        metadata_file = member
-            if not metadata_file:
-                msg = _('The archive at %(path)s does not contain a PKG-INFO file.')
-                msg = msg % {'path': archive_path}
-                raise ValueError(msg)
-
-            metadata_file_name = metadata_file.name
-            metadata_file = package_archive.extractfile(metadata_file)
-            metadata = metadata_file.read()
-
-            # Build a list of tuples of all the attributes found in the metadata. Ignore attributes
-            # with a leading underscore, as they are not part of the metadata.
-            attrs = dict()
-            try:
-                for attr in REQUIRED_ATTRS:
-                    attrs[attr] = re.search('^%s: (?P<field>.*?)\s*$' % cls._metadata_label(attr),
-                                            metadata, flags=re.MULTILINE).group('field')
-            except AttributeError:
-                msg = _('The PKG-INFO file is missing required attributes. Please ensure that the '
-                        'following attributes are all present: %(attrs)s')
-                msg = msg % {
-                    'attrs': ', '.join([cls._metadata_label(attr) for attr in REQUIRED_ATTRS])}
-                raise ValueError(msg)
-
-            # Add the filename, checksum, and checksum_type to the attrs
-            attrs['_filename'] = '%s-%s.tar%s' % (attrs['name'], attrs['version'], compression_type)
-            attrs['_checksum'] = checksum
-            attrs['_checksum_type'] = DEFAULT_CHECKSUM_TYPE
-            attrs['_metadata_file'] = metadata_file_name
-            package = cls(**attrs)
-            return package
-        finally:
-            if 'package_archive' in locals():
-                package_archive.close()
+    def from_json(cls, package_data, release, project_data):
```

Comment:
If/when PEP 426 is accepted, my interpretation is that the JSON served by a package index implementing the 2.0 version of the metadata should match the defined schema[0]. It might be worth leaving a note on this method that one day it might want to validate against that schema.

[0] https://www.python.org/dev/peps/pep-0426/#metadata-validation